### DISCUSSION TITLE
请问能否提供kook的java版示例？

### DISCUSSION THREAD

**Discussion by eden214** [simple-robot/Discussions#124] [Q&A] [answered]
### 涉及的编程语言

Java

### 涉及的组件库

kook

### 疑问描述

请问能否提供kook的java版示例？照着kook sdk首页和simbot官方文档研究半天也没研究出来java该如何使用

**Comment by ForteScarlet** [comment]
今晚会稍微整理一下KOOK组件的网站并提供一些快速开始的示例代码

**Comment by ForteScarlet** [comment, marked as answer]
https://simple-robot.github.io/simbot-component-kook/ 

可以试试看这里的**快速开始**内相关章节

**Reply by eden214** [reply]
> https://simple-robot.github.io/simbot-component-kook/
> 
> 可以试试看这里的**快速开始**内相关章节

我尝试了一下，有几个小问题：
1.看起来api只能使用ktor client，那创建client部分是否应该对使用者透明呢？
2.这部分的示例代码看起来跟github上的那个注册bot的风格不太一样，不确定是不是因为一个是ws一个是http
3.这部分示例代码的风格还是偏kotlin写法，对只用java的有点不太友好。。。

**Reply by ForteScarlet** [reply]
> 1. 看起来api只能使用ktor client，那创建client部分是否应该对使用者透明呢？

- `Ktor` 有关的配置可能会在后期优化，比如 https://github.com/simple-robot/simbot-component-kook/issues/91 以及自定义的引擎配置等。
这些应该会伴随着 https://github.com/simple-robot/simbot-component-kook/issues/82 完成

> 2. 这部分的示例代码看起来跟github上的那个注册bot的风格不太一样，不确定是不是因为一个是ws一个是http

- 目前只支持 `ws` 协议的"连接"，不知道你说的 "注册bot的风格" 是哪一种，因为网站中目前仅写了 "API模块" 和 "SpringBoot模块" 两种方式，那么你说的可能是使用 "标准库" 或 "simbot核心库" 的情况。
对于Java来说我更建议使用 SpringBoot 模块，因为它可以使用注解开发，会大大减少“对Java不友好”的情况，也就是第 `3` 条的情况。

> 3. 这部分示例代码的风格还是偏kotlin写法，对只用java的有点不太友好。。。

- 如上所述，在非注解开发模式的情况下，simbot是提供 Kotlin DSL 风格的API，而Java对于DSL来说本身就不是很友好，这难以避免，也因此更建议Java使用SpringBoot模块来减少这种情况。
除了 SpringBoot 模块以外，API模块、stdlib模块（标准库模块）是对KOOK API的**底层封装**，功能性封装较少，Java用起来稍微有点儿卡手也是无可奈何；而 simbot核心库模块则是"主要"面向Kotlin的模块，虽然依旧兼容Java，但是如你所见，DSL风格的API会降低Java的友好度。（没办法，Java的Lambda就是那种风格）

其实如果使用 simbot核心库模块 或者 SpringBoot模块，则**大部分**时候就不需要考虑底层API的问题了，因为这两个模块属于高一层的封装，大部分功能会屏蔽API的调用，比如发送消息可以直接 `sendBlocking("消息")` 而不是去使用 `MessageCreateRequest`。

如果真的希望不使用SpringBoot，而是坚持DSL风格的API，使用代码注册bot（比如 `new Bot(ticket,token).start()`）那我会考虑这两天补全一下剩下两个模块的快速开始

如果希望不使用SpringBoot，但是希望使用注解风格的API，也有办法，有一个带有轻量级IOC的简易模块，但是其不如Spring来的健壮和高级。

**Reply by ForteScarlet** [reply]
主要是之前一阵子都在重构 [QQ频道组件](https://github.com/simple-robot/simbot-component-qq-guild)，KOOK一直没顾上，还有一些小问题很正常。QQ频道组件目前的样子就大概率是KOOK组件未来的样子。

**Reply by eden214** [reply]
2.指的是https://github.com/simple-robot/simbot-component-kook  中简单完整示例部分注册bot监听的方式，这种的看起来是貌似不是springboot的，应该是基于核心库的？我本身对于直接使用api并没有什么太大的执念，直接使用核心库或springboot也是ok的。

最上面的问题中提到的kook示例其实指的就是这个简单完整示例，之前也是在尝试把简单完整示例转换为java写法，但是没有成功。。。请问核心库能提供示例吗？

另外如果核心库对java也不太友好的话，使用springboot会更好一些吗

**Reply by ForteScarlet** [reply]
哦，那个简单完整示例是使用 simbot核心库 的例子，如果用Java效果可以参考 https://simbot.forte.love/docs/quick-start/core

如果没什么特殊要求，用SpringBoot更好，毕竟注解在Java中比DSL要舒服一点儿

QQ频道和KOOK组件都还在 ALPHA 阶段，如果有什么非常严重的问题出现也可以考虑加QQ频道或者QQ群

**Reply by eden214** [reply]
非常感谢大佬解答~按照springboot的教程尝试是可以正常跑起示例了，但也遇到了一些新的困惑
1.目前springboot版本是没有自动心跳包保活的机制吗？还是需要怎么设置呢？
2.我是想搞点歌机器人，但目前kook官方文档并没有公布语音相关的api，所以是找了其他开发者获取到的接口，需要自己手动创建http请求用gatewayurl换语音频道的wss地址，然后ws连接语音频道推流，看目前kook stdlib的api，是否应该使用BotRequestUtil.requestDataBlocking()方法？还是其他方法呢？(貌似java只能引用这个方法和requestBlocking方法)如果是的话这里引入了3的问题。。。另外貌似也没看到支持创建ws的接口，还是说我应该通过其他http和ws的client自己手动连接，而不是通过kook sdk呢？
3.如果是需要用BotRequestUtil.requestDataBlocking()方法，这个方法貌似需要一个KookBot对象，我该如何获取呢？另外在springboot环境中，有没有一个currentBot的概念呢？
4.哪里有可以加入qq群和频道的地址呢？

**Reply by ForteScarlet** [reply]
现在在外面，先说问题4吧，组织库首页 https://github.com/simple-robot 的介绍里有群号或频道链接

**Reply by ForteScarlet** [reply]
> **1. 目前springboot版本是没有自动心跳包保活的机制吗？还是需要怎么设置呢？**

- 理论上是有的，如果遇到了bot因为心跳问题失活断连、可重连情况下无法重连等问题可到 [Issues](https://github.com/simple-robot/simpler-robot/issues/new?assignees=ForteScarlet&labels=bug%3F&template=bug-report.yml) 反馈。内部确实还有一些没处理好的细节需要继续打磨，也没有进行长时间的稳定性测试（只试过持续运行4个小时左右）

> **2. 如何获取 bot 当前 wss地址；3. 如何获取 `KookBot`**

- `KookBot` 的一些属性被隐藏了，不过可以考虑进行公开。后续版本会考虑开放相关属性，可追踪 https://github.com/simple-robot/simbot-component-kook/issues/94

- 使用 SpringBoot 时，你监听的每个事件中的 `.getBot` 均可能为 `KookComponentBot`，在使用前判断类型：
  ```java
  var bot = event.getBot();
  if (bot instanceof KookComponentBot kookCpBot) {
     kookCpBot.getSourceBot(); // 这就是 KookBot
  }
  ```
  或者直接监听KOOK组件的专属事件类型，比如 `KookChannelMessageEvent` ，此时 `event.getBot` 本身就是 `KookComponentBot` （不是那就是bug，如果被发现或反馈会修复）
  其次，有一个同名不同包的 `love.forte.simbot.component.kook.util.BotRequestUtil` 可以直接使用  `KookComponentBot` 来进行请求。 
  因为 `KookComponentBot` 是 **"Kook组件"** 中的bot实现，它针对 `stdlib` 中的 `KookBot` 封装了一层。

  > **Warning**
  > 遥远的未来重构之后 `KookComponentBot` 会改名为 `KookBot`，原本 stdlib 中的 `KookBot` 会直接命名为 `Bot`。

> **3. 创建ws的接口？**

- 针对**事件订阅**的ws连接创建是基于 `Ktor` 的，并且仅在内部使用，不对外公开。如果有什么其他创建websocket的需求，需要自行实现。

**Reply by ForteScarlet** [reply]
如果不需要一定获取 bot 当前的 ws 地址的话也可以考虑使用 `GatewayRequest` 主动获取新的 gateway 信息。

**Reply by eden214** [reply]
可能我没有表达清楚，不需要使用现有的ws连接，是需要使用发起新的http和ws请求（需要自己指定地址和一些参数，但token之类的信息是需要用现有bot的），想问问是否是现有的功能，如果没有的话我自行通过其他方式实现也可以

**Reply by ForteScarlet** [reply]
那就自行实现即可。
bot的token可以通过 `KookBot.getTicket` 获取，比如
```java
KookComponentBot bot = ...;
String token = bot.getSourceBot().getTicket().getToken(); // bot配置的token
```

**Reply by eden214** [reply]
好的，感谢~我留意到目前实现的api貌似并不完整，比如[频道用户相关接口列表]这个就没有，没有实现的部分目前也是需要自己手动处理吗？

**Reply by ForteScarlet** [reply]
目前API的确不是非常完整的，这些日后会逐渐补全。目前没有的实现确实需要临时的自行处理。

_当然也可以根据其他API实现一个临时的API类型，但是 Java 中这么做还是困难了点儿。_
